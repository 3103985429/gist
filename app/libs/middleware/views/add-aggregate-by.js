const winston = require('winston');
const { defaultTo } = require('lodash');

const { AGGREGATE_BY_DEFAULT_COLUMN_NAME } = require('../../../controllers/client/config');
const { fieldOverrideIfExists } = require('../../../controllers/client/func');
const {
    RealColumnNameFromHumanReadableColumnName,
    HumanReadableFEVisibleColumnNames_orderedForDropdown,
} = require('../../../libs/datasources/imported_data_preparation');
const { getCoercionOperations } = require('./get-coercion-operations');

module.exports = { addAggregateBy };

/**
 * aggregate by object has six fields:
 * - aggregateBy = the aggregate by field as supplied in the url query (either directly or via shared page). Will be
 * undefined if not explicitly specified, by design.
 * - aggregateBy_humanReadable = the aggregate column name that may or may not have been modified
 * - aggregateBy_realColumnName = the aggregate column name in its raw form
 * - colNames_orderedForAggregateByDropdown = column names to populate the dropdown
 * - defaultAggregateByColumnName = the default aggregate column name in its raw form
 * - defaultAggregateByColumnName_humanReadable = the default aggregate column name
 * @param {Object} req
 * @param {Object} res
 * @param {Function} next
 */
function addAggregateBy(req, res, next) {
    const {
        dataSource,
        query: { aggregateBy: queryAggregateBy },
        camelCaseViewType: viewType,
        viewSettings: masterViewSettings,
    } = req;
    const {
        defaultAggregateByColumnName,
        fieldsNotAvailableAsAggregateByColumns,
    } = dataSource.fe_views.views[viewType];
    const unavailableFields = defaultTo(fieldsNotAvailableAsAggregateByColumns, []);

    const aggregateBy = {
        aggregateBy: undefined,
    };

    if (queryAggregateBy) {
        aggregateBy.aggregateBy = queryAggregateBy;
    } else {
        aggregateBy.defaultAggregateByColumnName = defaultTo(defaultAggregateByColumnName, AGGREGATE_BY_DEFAULT_COLUMN_NAME);
    }

    if (!queryAggregateBy && defaultAggregateByColumnName) {
        aggregateBy.aggregateBy_humanReadable = fieldOverrideIfExists(aggregateBy.defaultAggregateByColumnName, dataSource);
        aggregateBy.aggregateBy_realColumnName = aggregateBy.defaultAggregateByColumnName;
    } else {
        aggregateBy.aggregateBy_humanReadable = defaultTo(aggregateBy.aggregateBy, aggregateBy.defaultAggregateByColumnName);
        aggregateBy.aggregateBy_realColumnName = defaultTo(RealColumnNameFromHumanReadableColumnName(aggregateBy.aggregateBy_humanReadable, dataSource), aggregateBy.aggregateBy_humanReadable);
    }

    aggregateBy.colNames_orderedForAggregateByDropdown = HumanReadableFEVisibleColumnNames_orderedForDropdown(dataSource, viewType, 'AggregateBy', getCoercionOperations('defaultAggregateByColumnName', masterViewSettings));

    if (!unavailableFields.includes(AGGREGATE_BY_DEFAULT_COLUMN_NAME)) {
        aggregateBy.colNames_orderedForAggregateByDropdown.push(AGGREGATE_BY_DEFAULT_COLUMN_NAME);
    }

    aggregateBy.defaultAggregateByColumnName_humanReadable = defaultTo(fieldOverrideIfExists(defaultAggregateByColumnName, dataSource), AGGREGATE_BY_DEFAULT_COLUMN_NAME);

    req.aggregateBy = aggregateBy;

    winston.debug('Added aggregateBy');
    next();
}
