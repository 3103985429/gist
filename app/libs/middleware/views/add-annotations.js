var winston = require('winston');

module.exports.addAnnotations = function (req, res, next) {
    var dataSourceDescription = req.dataSource;
    var type = req.camelCaseViewType;

    req.annotations = dataSourceDescription.fe_views.views[type].annotations;

    winston.debug('added annotations');
    next();
};
