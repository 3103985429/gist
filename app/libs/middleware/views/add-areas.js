var winston = require('winston');
var _ = require('lodash');

const { fieldOverrideIfExists } = require('../../../controllers/client/func');

module.exports.addAreas = function (req, res, next) {
    var dataSourceDescription = req.dataSource;
    var type = req.camelCaseViewType;
    var areaTitles = [];

    _.each(dataSourceDescription.fe_views.views[type].areas, function (area) {
        areaTitles.push({
            realName: area,
            humanReadable: fieldOverrideIfExists(area, dataSourceDescription),
        });
    });

    req.areas = areaTitles;
    req.simpleChart = dataSourceDescription.fe_views.views[type].simpleChart;

    winston.debug('added areas');
    next();
};
