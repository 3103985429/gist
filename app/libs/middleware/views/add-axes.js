const winston = require('winston');
const importedDataPreparation = require('../../../libs/datasources/imported_data_preparation');
const getCoercionOperations = require('./get-coercion-operations').getCoercionOperations;
const fieldOverrideIfExists = require('../../../controllers/client/func').fieldOverrideIfExists;

module.exports.addAxes = function (req, res, next) {
    const dataSourceDescription = req.dataSource;
    const query = req.query;
    const type = req.camelCaseViewType;
    const masterViewSettings = req.viewSettings;

    const axes = {
        xAxis: undefined,
        yAxis: undefined
    };

    // yAxis
    const yAxis = query.yAxis;
    axes.yAxis_humanReadable = fieldOverrideIfExists(yAxis || dataSourceDescription.fe_views.views[type].defaultYAxisField, dataSourceDescription);
    axes.yAxis_realName = yAxis ? importedDataPreparation.RealColumnNameFromHumanReadableColumnName(yAxis, dataSourceDescription) : (typeof dataSourceDescription.fe_views.views[type].defaultYAxisField === 'undefined') ? importedDataPreparation.RealColumnNameFromHumanReadableColumnName(axes.yAxis_humanReadable, dataSourceDescription) : dataSourceDescription.fe_views.views[type].defaultYAxisField;

    // xAxis
    const xAxis = query.xAxis;
    axes.xAxis_humanReadable = fieldOverrideIfExists(xAxis || dataSourceDescription.fe_views.views[type].defaultXAxisField, dataSourceDescription);
    axes.xAxis_realName = xAxis ? importedDataPreparation.RealColumnNameFromHumanReadableColumnName(xAxis, dataSourceDescription) : (typeof dataSourceDescription.fe_views.views[type].defaultXAxisField === 'undefined') ? importedDataPreparation.RealColumnNameFromHumanReadableColumnName(axes.xAxis_humanReadable, dataSourceDescription) : dataSourceDescription.fe_views.views[type].defaultXAxisField;

    axes.colNames_orderedForXAxisDropdown = importedDataPreparation.HumanReadableFEVisibleColumnNames_orderedForDropdown(dataSourceDescription, type, 'defaultXAxisField', getCoercionOperations('defaultXAxisField', masterViewSettings));
    axes.colNames_orderedForYAxisDropdown = importedDataPreparation.HumanReadableFEVisibleColumnNames_orderedForDropdown(dataSourceDescription, type, 'defaultYAxisField', getCoercionOperations('defaultYAxisField', masterViewSettings));

    axes.defaultXAxisField = dataSourceDescription.fe_views.views[type].defaultXAxisField;
    axes.defaultYAxisField = dataSourceDescription.fe_views.views[type].defaultYAxisField;

    // keep these undefined
    if (xAxis) {
        axes.xAxis = xAxis;
    }
    if (yAxis) {
        axes.yAxis = yAxis;
    }

    req.axes = axes;

    winston.debug('added axes');
    next();
};
