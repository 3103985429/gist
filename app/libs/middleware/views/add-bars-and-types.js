var winston = require('winston');
var _ = require('lodash');

const { fieldOverrideIfExists } = require('../../../controllers/client/func');

module.exports.addBarsAndType = function (req, res, next) {
    var dataSourceDescription = req.dataSource;
    var type = req.camelCaseViewType;
    var barTitles = [];

    let colFilterVals;
    if (req.colFilter && req.colFilter.col) {
        try {
            const vals = JSON.parse(req.colFilter.col);
            if (Array.isArray(vals)) {
                colFilterVals = vals;
            }
        } catch (e) {
            colFilterVals = [req.colFilter.col];
        }
    }

    _.each(dataSourceDescription.fe_views.views[type].bars, function (bar) {
        // only add bars that are published as visible
        if (dataSourceDescription.fe_excludeFields[bar] === false && (!colFilterVals || colFilterVals.some(col => col === bar))) {
            barTitles.push({
                realName: bar,
                humanReadable: fieldOverrideIfExists(bar, dataSourceDescription),
            });
        }
    });

    req.bars = {
        bars: barTitles,
        type: dataSourceDescription.fe_views.views[type].groupedBars ? 'grouped' : 'stacked',
    };
    req.simpleChart = dataSourceDescription.fe_views.views[type].simpleChart;

    winston.debug('added bars and type');
    next();
};
