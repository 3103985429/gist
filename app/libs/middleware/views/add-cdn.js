const winston = require('winston');
const { defaultTo } = require('lodash');


module.exports.addCdn = function (req, res, next) {
    const { cdnAddress } = req.query;

    req.cdnAddress = defaultTo(cdnAddress, '');

    winston.debug('added cdn');
    next();
};
