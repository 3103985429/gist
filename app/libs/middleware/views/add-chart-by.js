var winston = require('winston');

const { fieldOverrideIfExists } = require('../../../controllers/client/func');
var importedDataPreparation = require('../../../libs/datasources/imported_data_preparation');
var getCoercionOperations = require('./get-coercion-operations').getCoercionOperations;

module.exports.addChartBy = function (req, res, next) {
    var dataSourceDescription = req.dataSource;
    var type = req.camelCaseViewType;
    var query = req.query;
    var masterViewSettings = req.viewSettings;

    var chartBy = {};

    chartBy.chartBy = query.chartBy; // the human readable col name - real col name derived below
    chartBy.defaultChartByColumnName_humanReadable = fieldOverrideIfExists(dataSourceDescription.fe_views.views[type].defaultChartByColumnName, dataSourceDescription);

    chartBy.colNames_orderedForChartByDropdown = importedDataPreparation.HumanReadableFEVisibleColumnNames_orderedForDropdown(dataSourceDescription, type, 'ChartBy', getCoercionOperations('defaultChartByColumnName', masterViewSettings));

    chartBy.chartBy_realColumnName = chartBy.chartBy ? importedDataPreparation.RealColumnNameFromHumanReadableColumnName(chartBy.chartBy, dataSourceDescription) :
        (dataSourceDescription.fe_views.views[type].defaultChartByColumnName === 'Object Title') ? importedDataPreparation.RealColumnNameFromHumanReadableColumnName(dataSourceDescription.fe_views.views[type].defaultChartByColumnName, dataSourceDescription) :
            dataSourceDescription.fe_views.views[type].defaultChartByColumnName;

    req.chartBy = chartBy;

    winston.debug('added chartBy');
    next();
};
