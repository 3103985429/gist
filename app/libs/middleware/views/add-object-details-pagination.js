const winston = require('winston');
const { isNil } = require('lodash');

const {
    VIEW_TYPE_GALLERY,
    VIEW_TYPE_TIMELINE,
    VIEW_TYPE_BUBBLE_CHART,
    VIEW_TYPE_SCATTERPLOT,
    VIEW_TYPE_TABLE,
    VIEW_TYPE_MAP,
    VIEW_TYPE_GLOBE,
    VIEW_TYPE_REGIONAL_MAP,
} = require('../../../config/views.config');
const { Lazy_Shared_ProcessedRowObject_MongooseContext } = require('../../../models/processed_row_objects');
const { getProject } = require('../../../controllers/client/data_preparation/data_prep_helpers/project.helpers');
const { getSort } = require('../../../controllers/client/data_preparation/data_prep_helpers/sort.helpers');
const { getMatch } = require('../../../controllers/client/data_preparation/data_prep_helpers/match.helpers');
const {
    addDefaultAggregationOperators,
    aggregateProcessedRowObjects,
} = require('../../../controllers/client/data_preparation/data_prep_helpers/aggregation.helpers');

module.exports = { addObjectDetailsPagination };

const temporarilyPaginatedViews = [
    VIEW_TYPE_GALLERY,
    VIEW_TYPE_TIMELINE,
    VIEW_TYPE_BUBBLE_CHART,
    VIEW_TYPE_SCATTERPLOT,
    VIEW_TYPE_TABLE,
    VIEW_TYPE_MAP,
    VIEW_TYPE_GLOBE,
    VIEW_TYPE_REGIONAL_MAP,
];

async function addObjectDetailsPagination(req, res, next) {
    const { dataSource, dataPrepOptions: options, viewType, query: { objectIndex } } = req;

    if (isNil(objectIndex) || !temporarilyPaginatedViews.includes(viewType)) {
        return next();
    }

    const parsedObjectIndex = Number.parseInt(objectIndex) || 0;
    const skip = Math.max(0, parsedObjectIndex - 1);
    const limit = 3;

    const { Model: processedRowObjectsModel } = Lazy_Shared_ProcessedRowObject_MongooseContext(dataSource._id);
    const aggregationOperators = addDefaultAggregationOperators(dataSource, options);
    const additionalProjects = getProject(viewType, dataSource, options);
    const preProjects = {
        $project: {
            _id: 1,
            ...additionalProjects,
        },
    };

    const postProjects = {
        $project: {
            id: '$_id',
            title: `$rowParams.${dataSource.objectTitle}`,
        },
    };
    const sort = getSort(viewType, dataSource, options);
    const match = getMatch(viewType, dataSource, options);
    const pagedAggregationOperators = aggregationOperators.concat([
        preProjects,
        { $match: match },
        { $sort: sort },
        { $skip: skip },
        { $limit: limit },
        postProjects,
    ]);
    const docs = await aggregateProcessedRowObjects(processedRowObjectsModel, pagedAggregationOperators);

    req.objectDetailsPagination = parsedObjectIndex === 0 ? { next: docs[1] } : { prev: docs[0], next: docs[2] };
    req.objectIndex = parsedObjectIndex;

    winston.debug('Added object details pagination');
    next();
}
