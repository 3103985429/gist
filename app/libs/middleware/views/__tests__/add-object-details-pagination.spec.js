import { pokemon1 } from '../../../../../internals/testing/backend/fixtures/datasets';
import { hydrateDb } from '../../../../../internals/testing/backend/utils';
import { VIEW_TYPE_AREA_CHART, VIEW_TYPE_GALLERY } from '../../../../config/views.config';
import { addObjectDetailsPagination } from '../add-object-details-pagination';


describe('Add object details pagination middleware', () => {
    let req, res, next;

    beforeEach(async () => {
        await hydrateDb();
        req = {
            query: {},
            dataSource: Object.assign({}, pokemon1),
            dataPrepOptions: {},
        };
        res = {};
        next = jest.fn();
    });

    it('should not add pagination when no object index is provided', async () => {
        await addObjectDetailsPagination(req, res, next);

        expect(req).not.toHaveProperty('objectIndex');
    });

    it('should not add pagination for not included view', async () => {
        req.query.viewType = VIEW_TYPE_AREA_CHART;
        await addObjectDetailsPagination(req, res, next);

        expect(req).not.toHaveProperty('objectIndex');
    });

    it('should add pagination for first result', async () => {
        req.viewType = VIEW_TYPE_GALLERY;
        req.query.objectIndex = 0;
        await addObjectDetailsPagination(req, res, next);

        expect(req).toHaveProperty('objectDetailsPagination', expect.any(Object));
        expect(req).toHaveProperty('objectIndex', 0);
        expect(req.objectDetailsPagination).toMatchSnapshot();
    });

    it('should add pagination for another result', async () => {
        req.viewType = VIEW_TYPE_GALLERY;
        req.query.objectIndex = 2;
        await addObjectDetailsPagination(req, res, next);

        expect(req).toHaveProperty('objectDetailsPagination', expect.any(Object));
        expect(req).toHaveProperty('objectIndex', 2);
        expect(req.objectDetailsPagination).toMatchSnapshot();
    });
});
