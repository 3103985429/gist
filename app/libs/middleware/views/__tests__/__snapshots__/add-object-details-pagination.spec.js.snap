// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add object details pagination middleware should add pagination for another result 1`] = `
Object {
  "next": Object {
    "_id": "5b05346ad7a86407e5ef783e",
    "id": "5b05346ad7a86407e5ef783e",
    "title": 2,
  },
  "prev": Object {
    "_id": "5b05346ad7a86407e5ef783d",
    "id": "5b05346ad7a86407e5ef783d",
    "title": 1,
  },
}
`;

exports[`Add object details pagination middleware should add pagination for first result 1`] = `
Object {
  "next": Object {
    "_id": "5b05346ad7a86407e5ef783d",
    "id": "5b05346ad7a86407e5ef783d",
    "title": 1,
  },
}
`;
